pkg/workflows/wasm/host/module.go

```diff
@@ -429,19 +424,177 @@ func fetchFn(logger logger.Logger, modCfg *ModuleConfig) func(caller *wasmtime.C
-		size = copyBuffer(caller, resplenBytes, resplenptr, uint32Size)
-		if size == -1 {
+		if size := wasmWriteUInt32(caller, resplenptr, uint32(len(respBytes))); size == -1 {
 			return ErrnoFault
 		}
 
 		return ErrnoSuccess
 	}
 }
+
+// createEmitFn injects dependencies and builds the emit function exposed by the WASM.  Errors in
+// Emit, if any, are returned in the Error Message of the response.
+func createEmitFn(
+	l logger.Logger,
+	e sdk.Emitter,
+	reader UnsafeReaderFunc,
+	writer UnsafeWriterFunc,
+	sizeWriter UnsafeFixedLengthWriterFunc,
+) func(caller *wasmtime.Caller, respptr, resplenptr, msgptr, msglen int32) int32 {
+	logErr := func(err error) {
+		l.Errorf("error emitting message: %s", err)
+	}
+
+	return func(caller *wasmtime.Caller, respptr, resplenptr, msgptr, msglen int32) int32 {
+		// writeErr marshals and writes an error response to wasm
+		writeErr := func(err error) int32 {
+			logErr(err)
+
+			resp := &wasmpb.EmitMessageResponse{
+				Error: &wasmpb.Error{
+					Message: err.Error(),
+				},
+			}
+
+			respBytes, perr := proto.Marshal(resp)
+			if perr != nil {
+				logErr(perr)
+				return ErrnoFault
+			}
+
+			if size := writer(caller, respBytes, respptr, int32(len(respBytes))); size == -1 {
+				logErr(errors.New("failed to write response"))
+				return ErrnoFault
+			}
+
+			if size := sizeWriter(caller, resplenptr, uint32(len(respBytes))); size == -1 {
+				logErr(errors.New("failed to write response length"))
+				return ErrnoFault
+			}
+
+			return ErrnoSuccess
+		}
+
+		b, err := reader(caller, msgptr, msglen)
+		if err != nil {
+			return writeErr(err)
+		}
+
+		msg := &wasmpb.EmitMessageRequest{}
+		err = proto.Unmarshal(b, msg)
+		if err != nil {
+			return writeErr(err)
+		}
+
+		vl, err := values.FromMapValueProto(msg.Labels)
+		if err != nil {
+			return writeErr(err)
+		}
+
+		// Handle the case of no labels before unwrapping.
+		if vl == nil {
+			vl = values.EmptyMap()
+		}
+
+		var labels map[string]any
+		if err := vl.UnwrapTo(&labels); err != nil {
+			return writeErr(err)
+		}
+
+		if err := e.Emit(msg.Message, labels); err != nil {
+			return writeErr(err)
+		}
+
+		return ErrnoSuccess
+	}
+}
+
+// beholderEmitter sends a message to the Beholder service via the custmsg package's labeler.  Each label is
+// expected to be a string.  Fails if any label is not a string.
+func beholderEmitter(msg string, labels map[string]any) error {
+	validated := make(map[string]string, len(labels))
+	for k, v := range labels {
+		str, ok := v.(string)
+		if !ok {
+			return fmt.Errorf("label %s is not a string", k)
+		}
+		validated[k] = str
+	}
+
+	return custmsg.NewLabeler().WithMapLabels(validated).SendLogAsCustomMessage(msg)
+}
+
+// UnsafeWriterFunc defines behavior for writing directly to wasm memory.  A source slice of bytes
+// is written to the location defined by the ptr.
+type UnsafeWriterFunc func(c *wasmtime.Caller, src []byte, ptr, len int32) int64
```

Comment:
@MStreet3 Can we make these types private? They shouldn't leak outside the package.